packages/backend/src/services/PromoteService/PromoteService.ts

```diff
@@ -0,0 +1,764 @@
+        }
+        const newSpace = await this.spaceModel.createSpace(
+            upstreamContent.projectUuid,
+            promotedContent.space.name,
+            user.userId,
+            promotedContent.space.isPrivate,
+            promotedContent.space.slug,
+        );
+        return { uuid: newSpace.uuid, isNew: true };
+    }
+
+    private async upsertChart(
+        user: SessionUser,
+        promotedContent: PromotedChart,
+        upstreamContent: UpstreamChart,
+    ) {
+        const upstreamChart = upstreamContent.chart;
+        const promotedChart = promotedContent.chart;
+        if (upstreamChart === undefined) {
+            // Create chart
+            if (promotedChart?.dashboardUuid) {
+                // Create chat in dashboard
+                return this.savedChartModel.create(
+                    upstreamContent.projectUuid,
+                    user.userUuid,
+                    {
+                        ...promotedChart,
+                        dashboardUuid: promotedChart.dashboardUuid,
+                        spaceUuid: undefined,
+                        updatedByUser: user,
+                        slug: promotedChart.slug,
+                    },
+                );
+            }
+
+            // Create chart in space
+            const space = await this.getOrCreateSpace(
+                user,
+                promotedContent,
+                upstreamContent,
+            );
+
+            return this.savedChartModel.create(
+                upstreamContent.projectUuid,
+                user.userUuid,
+                {
+                    ...promotedChart,
+                    dashboardUuid: undefined,
+                    spaceUuid: space.uuid,
+                    updatedByUser: user,
+                    slug: promotedChart.slug,
+                },
+            );
+        }
+
+        if (
+            upstreamChart.name !== promotedChart.name ||
+            upstreamChart.description !== promotedChart.description ||
+            upstreamChart.spaceUuid !== upstreamContent.space?.uuid
+        ) {
+            // We also update chart name and description if they have changed
+            await this.savedChartModel.update(upstreamChart.uuid, {
+                name: promotedChart.name,
+                description: promotedChart.description,
+                spaceUuid: upstreamContent.space?.uuid,
+            });
+        }
+        // update chart
+        const updatedChart = await this.savedChartModel.createVersion(
+            upstreamChart.uuid,
+            {
+                tableName: promotedChart.tableName,
+                metricQuery: promotedChart.metricQuery,
+                chartConfig: promotedChart.chartConfig,
+                tableConfig: promotedChart.tableConfig,
+            },
+            user,
+        );
+        return updatedChart;
+    }
+
+    async promoteChart(user: SessionUser, chartUuid: string) {
```

Comment:
This is the main method to promote chart, compared to the old one on projectService, this one is more modular and readable